SetupDataPkg/ConfProfileMgrDxe/ConfProfileMgrDxe.c

```diff
@@ -0,0 +1,248 @@
+    // not matching the variable store
+    if ((Status != EFI_BUFFER_TOO_SMALL) && EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_ERROR, "%a failed to read variable %s Status: (%r)!\n", __FUNCTION__, VarList[i].Name, Status));
+      ASSERT (FALSE);
+      ValidationFailure = TRUE;
+      goto Done;
+    }
+
+    if ((Size != VarList[i].DataSize) ||
+        (0 != CompareMem (Data, VarList[i].Data, Size)) ||
+        (Attributes != VarList[i].Attributes))
+    {
+      DEBUG ((DEBUG_ERROR, "%a active profile does not match variable store %s!\n", __FUNCTION__, VarList[i].Name));
+      ValidationFailure = TRUE;
+      goto Done;
+    }
+
+    FreePool (Data);
+    Data = NULL;
+  }
+
+Done:
+  if (Data != NULL) {
+    FreePool (Data);
+  }
+
+  if (ValidationFailure) {
+    // Write profile values and reset the system
+    for (i = 0; i < VarListCount; i++) {
+      Status = gRT->SetVariable (
+                      VarList[i].Name,
+                      &VarList[i].Guid,
+                      VarList[i].Attributes,
+                      VarList[i].DataSize,
+                      VarList[i].Data
+                      );
+
+      // We should not fail to write any of these variables, but if so, try to write the rest
+      if (EFI_ERROR (Status)) {
+        DEBUG ((DEBUG_ERROR, "%a failed to write variable %s Status: (%r)!\n", __FUNCTION__, VarList[i].Name, Status));
+        ASSERT (FALSE);
+      }
+    }
+
+    DEBUG ((DEBUG_ERROR, "%a profile written, resetting system\n", __FUNCTION__));
+    ResetSystemWithSubtype (EfiResetCold, &gConfProfileMgrResetGuid);
+    // Reset system should not return, dead loop if it does
+    CpuDeadLoop ();
+  }
+
+  if (VarList != NULL) {
+    for (i = 0; i < VarListCount; i++) {
+      // Also free the data and name
+      if (VarList[i].Name != NULL) {
+        FreePool (VarList[i].Name);
+      }
+
+      if (VarList[i].Data != NULL) {
+        FreePool (VarList[i].Data);
+      }
+    }
+
+    FreePool (VarList);
+  }
+
+  return;
+}
+
+/**
+  This driver will query the active configuration profile and validate the flash contents match. If not, the
+  profile contents will be written to flash and the system reset.
+
+  @param  ImageHandle  Handle to this image.
+  @param  SystemTable  Pointer to the system table
+
+  @retval  EFI_SUCCESS   The driver validated the profile against flash contents successfully.
+  @retval  !EFI_SUCCESS  The driver failed to validate the profile against flash contents successfully.
+**/
+EFI_STATUS
+EFIAPI
+ConfProfileMgrDxeEntry (
+  IN  EFI_HANDLE        ImageHandle,
+  IN  EFI_SYSTEM_TABLE  *SystemTable
+  )
+{
+  EFI_GUID    *ActiveProfileGuid = NULL;
+  EFI_STATUS  Status;
+  UINTN       Size = sizeof (EFI_GUID);
+  UINT32      Attributes;
+  EFI_GUID    CachedProfile;
+  BOOLEAN     NeedToFreeMem = TRUE;
+
+  // RetrieveActiveProfileGuid will validate the profile guid is valid before returning
+  Status = RetrieveActiveProfileGuid (&ActiveProfileGuid);

```

Comment:
The function call returns a guid.

Could we get rid of the need to allocate/free memory?

I think the code in this function could be simplified by getting rid of the allocation and tracking of allocation. 

If we eliminate the allocating of memory, I think the ActiveProfileGuid can be removed too, and the logic can be simplieied to 
try to get activeprofileguid into local variable
try to get cached profileguid to local variable
copy gSetupDataPkgGenericProfileGuid into local variable.